### ISSUE TITLE
Jadd_*() segfaults if JSON object is not json_type_object

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#181]
`t/lua/t0002-rpc.t` sends a ping request with a JSON array payload.

If there weren't an explicit test in the ping handler to catch this and return EPROTO, a subsequent Jadd_str() in the handler would segfault.  The `Jadd_*()` functions return void.

Note that RFC 5 decrees that "lsmod" shall return a JSON array, so we have documented protocol where a JSON payload is not `json_type_object`

This is a segfault waiting to happen in most of our RPC handlers.

**Comment by trws** [comment]
Where is the resulting segfault? Is this a result of a check in the RPC/request code or what is it?  It doesn't seem immediately clear from the report.

**Comment by garlick** [comment]
The segfault occurs in `json_object_object_add()` its first argument a type other than json_type_object.

The `shortjson.h::Jadd_*()` functions could get an assertion and some documentation:
```
assert (json_object_is_type (o, json_type_object));
```

**Comment by grondo** [comment]
Is there a bug in the `lua/t0002-rpc.t` test that needs to be resolved?

**Comment by garlick** [comment]
No problem with the test which tries to elicit an EPROTO by sending a ping request with an array payload.

The ping handler makes a copy of the payload to echo back, and modifies it with `Jadd_str()` to return the route.  The test would elicit a segfault instead of an EPROTO without a check that the received object is of type `json_type_object`.  This points to a general fragility of `Jadd_*()` and that is what the issue is about.

**Comment by trws** [comment]
Do we have a specification for the json that is valid inside a message? Base object arrays, ints, strings are all technically valid json, but maybe it would be worth specifying that an object is the base unit, and everything must be specified as keys on that base object?

**Comment by garlick** [comment]
That's not a bad thought.  Perhaps we could add a section to RFC 3 on valid JSON content.  That would enable us to provide a decode function that throws EPROTO on invalid content, without having to replicate this in every handler.

We would have to update the lsmod protocol defined in RFC 5, which uses a JSON array as the outermost type in the payload.

**Comment by garlick** [comment]
I got back to this when attempting to work around #638 by removing the `live` module from the rc1 default module list.  I updated the aforementioned lua rpc test to ping the broker rather than the live module, and as the broker lacked the explicit check for a non-array json object, it failed.

Let's go ahead and restrict JSON payloads to object only.   This should be done in the low level message encode/decode functions, and then we will need to update RFC 3.

We will then need to update lsmod and RFC 5 (as mentioned above).

We will also need to fix the live "hello" protocol, which is not yet documented in an RFC.
